release-20.1: backupccl: stop blocking backup jobs on client disconnect

pbardea opened the pull request:
Backport 1/1 commits from #51999.

/cc @cockroachdb/release

---

Previously, a backup/restore job whose client disconnect might have blocked on
sending its results to the resultCh (and thus block forever). This
commit allows these jobs to complete.

Release note (bug fix): Previously, a BACKUP/RESTORE jobs would block once it
finished performing the job. This is now fixed.

pbardea committed:
backupccl: stop blocking backup jobs on client disconnect

Previously, a backup job whose client disconnect might have blocked on
sending its results to the resultCh (and thus block forever). This
commit allows these jobs to complete.

Release note (bug fix): Previously, a BACKUP job would block once it
finished backing up the data.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52003)
<!-- Reviewable:end -->

pbardea committed:
backupccl: don't block RESTORE when client disconnects

This commit prevents RESTORE from blocking when sending its results back
over resultCh if the client that started the job disconnects.

Release note (bug fix): Previously, RESTORE would sometimes block at the
end of the job when sending its results back if the connection that
started the job disconnected. This is now fixed.
